spec/src/main/asciidoc/core/invokers.asciidoc

```diff
@@ -0,0 +1,181 @@
+----
+public interface Invoker<T, R> {
+    R invoke(T instance, Object[] arguments);
+}
+----
+
+This section describes the behavior of the `invoke()` method.
+
+Calling `invoke()` invokes the target method on given `instance` of the target bean, passing given `arguments`.
+
+If the target method is `static`, the `instance` is ignored; by convention, it should be `null`.
+If the target method returns normally, its return value is returned, unless the target method is declared `void`, in which case `null` is returned.
+If the target method throws an exception, it is rethrown.
+The `invoke()` method does not declare any checked exception; implementations are supposed to use the "sneaky throw" idiom.
+// TODO maybe change that?
+
+If the target method is not `static` and `instance` is `null`, a `NullPointerException` is thrown.
+If the target method is not `static` and the `instance` is not <<invoker_assignability,assignable>> to the target bean, a `ClassCastException` is thrown.
+
+Correspondence between given `arguments` and declared parameters of the target method is positional: the Nth element of the `arguments` array is passed as the Nth argument to the target method.
+If the target method is a variable arity method, the last element of the `arguments` array corresponds to the variable arity parameter (and therefore must be an array).
+
+If the target method declares no parameter, `arguments` are ignored.
+If the target method declares any parameter and `arguments` is `null`, `NullPointerException` is thrown.
+If the `arguments` array has fewer elements than the number of parameters of the target method, `ArrayIndexOutOfBoundsException` is thrown.
+If the `arguments` array has more elements than the number of parameters of the target method, the excess elements are ignored.
+If some of the `arguments` is not <<invoker_assignability,assignable>> to the corresponding parameter of the target method, `ClassCastException` is thrown.
+
+If the declared type of a parameter of the target method is a primitive type, then:
+
+* if the corresponding argument is `null`, the default value of the primitive type is passed;
+* if the corresponding argument is not `null`, unboxing conversion is performed on the argument and the result is passed.
+
+If the declared type of a parameter of the target method is not a primitive type, the corresponding argument is passed as is.
+
+// TODO when the `InvokerBuilder` applies transformations, some of the requirements above are no longer strictly necessary, we should reflect that in this text somehow
+
+Invoker implementations must be thread-safe.
+A single invoker instance may be used to perform multiple, possibly concurrent, invocations of the target method, possibly on different instances of the target bean, possibly with different arguments.
+
+Whenever a direct invocation of a method is a business method invocation, an indirect invocation of that method through an invoker is also a business method invocation.
```

Comment:
This is one of the key behaviors; I'd move it all the way to the `== Method invokers` section.

Reply:
Agree it is a little hidden here. I'll move it.

Reply:
I moved this up in the "Using an `Invoker` chapter, above the newly added subsection that specifies all the boring but necessary details.